File: server/common/src/main/java/su/plo/voice/server/player/LuckPermsListener.java
package su.plo.voice.server.player;

import com.google.common.collect.Maps;
import lombok.RequiredArgsConstructor;
import net.luckperms.api.LuckPerms;
import net.luckperms.api.LuckPermsProvider;
import net.luckperms.api.event.EventBus;
import net.luckperms.api.event.EventSubscription;
import net.luckperms.api.event.node.NodeAddEvent;
import net.luckperms.api.event.node.NodeClearEvent;
import net.luckperms.api.event.node.NodeMutateEvent;
import net.luckperms.api.event.node.NodeRemoveEvent;
import net.luckperms.api.model.group.Group;
import net.luckperms.api.model.user.User;
import net.luckperms.api.node.Node;
import org.jetbrains.annotations.NotNull;
import su.plo.voice.api.server.PlasmoVoiceServer;
import su.plo.voice.api.server.player.VoicePlayer;
import su.plo.voice.api.server.player.VoicePlayerManager;
import su.plo.voice.server.event.player.PlayerPermissionUpdateEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

@RequiredArgsConstructor
public final class LuckPermsListener {

    private final PlasmoVoiceServer voiceServer;
    private final VoicePlayerManager players;
    private final ScheduledExecutorService executor;

    private final LuckPerms luckPerms = LuckPermsProvider.get();
    private final List<EventSubscription<?>> subscriptions = new ArrayList<>();

    private final Map<String, ScheduledFuture<?>> permissionChanges = Maps.newHashMap();

    public void subscribe() {
        EventBus bus = luckPerms.getEventBus();

        EventSubscription<?> subscription = bus.subscribe(NodeAddEvent.class, this::onNodeAdd);
        subscriptions.add(subscription);

        subscription = bus.subscribe(NodeRemoveEvent.class, this::onNodeRemove);
        subscriptions.add(subscription);

        subscription = bus.subscribe(NodeClearEvent.class, this::onNodeClear);
        subscriptions.add(subscription);
    }

    public void unsubscribe() {
        subscriptions.forEach(EventSubscription::close);
    }

    private void onNodeAdd(NodeAddEvent event) {
        onNodeMutate(event, event.getNode());
    }

    private void onNodeRemove(NodeRemoveEvent event) {
        onNodeMutate(event, event.getNode());
    }

    private void onNodeClear(NodeClearEvent event) {
        for (Node node : event.getNodes()) {
            onNodeMutate(event, node);
        }
    }

    private void onNodeMutate(NodeMutateEvent event, Node node) {
        if (event.isUser()) {
            User user = (User) event.getTarget();
            Optional<VoicePlayer> player = players.getPlayerById(user.getUniqueId());
            if (!player.isPresent() || !player.get().hasVoiceChat()) return;

            onLpPermissionChange(player.get(), node.getKey());
        } else if (event.isGroup()) {
            Group group = (Group) event.getTarget();

            // iterate all online players and check if they are in group
            for (VoicePlayer player : players.getPlayers()) {
                if (player.getInstance().hasPermission("group." + group.getName())) {
                    onLpPermissionChange(player, node.getKey());
                }
            }
        }
    }

    private synchronized void onLpPermissionChange(@NotNull VoicePlayer player, @NotNull String permission) {
        String playerPermissionKey = player.getInstance().getUUID() + "_" + permission;

        ScheduledFuture<?> future = permissionChanges.get(playerPermissionKey);
        if (future != null) future.cancel(false);

        permissionChanges.put(
                player.getInstance().getUUID() + "_" + permission,
                executor.schedule(
                        () -> onPermissionChange(player, permission),
                        100L,
                        TimeUnit.MILLISECONDS
                )
        );
    }

    private void onPermissionChange(@NotNull VoicePlayer player, @NotNull String permission) {
        System.out.println(permission);
        voiceServer.getEventBus().call(new PlayerPermissionUpdateEvent(player, permission));
        permissionChanges.remove(player.getInstance().getUUID() + "_" + permission);
    }
}

File: server-common/src/main/kotlin/su/plo/voice/server/player/LuckPermsListener.kt
package su.plo.voice.server.player

import com.google.common.collect.Maps
import net.luckperms.api.LuckPerms
import net.luckperms.api.LuckPermsProvider
import net.luckperms.api.event.EventSubscription
import net.luckperms.api.event.node.NodeAddEvent
import net.luckperms.api.event.node.NodeClearEvent
import net.luckperms.api.event.node.NodeMutateEvent
import net.luckperms.api.event.node.NodeRemoveEvent
import net.luckperms.api.event.user.UserDataRecalculateEvent
import net.luckperms.api.model.group.Group
import net.luckperms.api.model.user.User
import net.luckperms.api.node.Node
import su.plo.voice.api.server.PlasmoBaseVoiceServer
import su.plo.voice.api.server.event.player.PlayerPermissionUpdateEvent
import su.plo.voice.api.server.player.VoicePlayer
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class LuckPermsListener(
    private val voiceServer: PlasmoBaseVoiceServer,
    private val executor: ScheduledExecutorService
) {

    private val luckPerms: LuckPerms = LuckPermsProvider.get()
    private val subscriptions: MutableList<EventSubscription<*>> = ArrayList()
    private val permissionChanges: MutableMap<String, ScheduledFuture<*>> = Maps.newHashMap()

    fun subscribe() {
        val luckPermsBus = luckPerms.eventBus

        subscriptions.add(luckPermsBus.subscribe(
            NodeAddEvent::class.java,
            ::onNodeAdd
        ))

        subscriptions.add(luckPermsBus.subscribe(
            NodeRemoveEvent::class.java,
            ::onNodeRemove
        ))

        subscriptions.add(luckPermsBus.subscribe(
            NodeClearEvent::class.java,
            ::onNodeClear
        ))
    }

    fun unsubscribe() {
        subscriptions.forEach { it.close() }
    }

    private fun onNodeAdd(event: NodeAddEvent) {
        onNodeMutate(event, event.node)
    }

    private fun onNodeRemove(event: NodeRemoveEvent) {
        onNodeMutate(event, event.node)
    }

    private fun onNodeClear(event: NodeClearEvent) {
        event.nodes.forEach {
            onNodeMutate(event, it)
        }
    }

    private fun onNodeMutate(event: NodeMutateEvent, node: Node) {
        if (event.isUser) {
            val user = event.target as User

            val player = voiceServer.playerManager.getPlayerById(user.uniqueId, false).orElse(null) ?: return
            if (!player.hasVoiceChat()) return

            onLpPermissionChange(player, node.key)
        } else if (event.isGroup) {
            val group = event.target as Group

            // iterate all online players and check if they are in group
            voiceServer.playerManager.players.forEach { player ->
                if (player.instance.hasPermission("group." + group.name)) {
                    onLpPermissionChange(player, node.key)
                }
            }
        }
    }

    @Synchronized
    private fun onLpPermissionChange(player: VoicePlayer, permission: String) {
        val playerPermissionKey: String = player.instance.uuid.toString() + "_" + permission
        val future = permissionChanges[playerPermissionKey]
        future?.cancel(false)

        permissionChanges[player.instance.uuid.toString() + "_" + permission] = executor.schedule(
            { onPermissionChange(player, permission) },
            100L,
            TimeUnit.MILLISECONDS
        )
    }

    private fun onPermissionChange(player: VoicePlayer, permission: String) {
        voiceServer.eventBus.call(PlayerPermissionUpdateEvent(player, permission))
        permissionChanges.remove(player.instance.uuid.toString() + "_" + permission)
    }

    companion object {

        fun hasLuckPerms() =
            try {
                Class.forName("net.luckperms.api.LuckPermsProvider")
                true
            } catch (_: Exception) {
                false
            }
    }
}
